### ISSUE TITLE
Conda install error, incomplete output from deepbgc pipeline, and IndexError: tuple index out of range

### ISSUE THREAD

**Issue by ZarulHanifah** [Merck/deepbgc#11]
Hello DeepBGC developers,

I learned about this software from Geoffrey Hannigan's recent microseminar - fantastic. I just had a quick test on the software. I tried _conda install_ the software but it didn't work. Conda output something like this:

```
UnsatisfiableError: The following specifications were found to be in conflict:
  - deepbgc
Use "conda search <package> --info" to see the dependencies for each package.
```

Anyways, I proceeded with creating an empty conda environment, conda install the dependencies and pip install deepbgc. DeepBGC worked so far, although there was a lot of deprecation warning from tensorflow. I downloaded the trained classifier with _deepbgc download_. Here is the output for _deepbgc info_:

```
 _____                  ____    ____   ____ 
 |  _ \  ___  ___ ____ | __ )  / ___) / ___)
 | | \ \/ _ \/ _ \  _ \|  _ \ | |  _ | |    
 | |_/ /  __/  __/ |_) | |_) || |_| || |___ 
 |____/ \___|\___| ___/|____/  \____| \____)
=================|_|===== version 0.1.7 =====
INFO    27/07 23:14:23   Available data files: ['Pfam-A.31.0.hmm', 'Pfam-A.31.0.hmm.h3m', 'Pfam-A.31.0.hmm.h3f', 'Pfam-A.31.0.hmm.h3i', 'Pfam-A.31.0.hmm.h3p', 'Pfam-A.31.0.clans.tsv']
INFO    27/07 23:14:23   ================================================================================
INFO    27/07 23:14:23   Available detectors: ['clusterfinder_geneborder', 'clusterfinder_retrained', 'deepbgc', 'clusterfinder_original']
INFO    27/07 23:14:23   --------------------------------------------------------------------------------
INFO    27/07 23:14:23   Model: clusterfinder_geneborder
INFO    27/07 23:14:23   Loading model from: /home/zarul/.local/share/deepbgc/data/0.1.0/detector/clusterfinder_geneborder.pkl
INFO    27/07 23:14:23   Type: GeneBorderHMM
INFO    27/07 23:14:23   Version: 0.1.0
INFO    27/07 23:14:23   Timestamp: 1551449863.1564941 (2019-03-01T22:17:43.156494)
INFO    27/07 23:14:23   --------------------------------------------------------------------------------
INFO    27/07 23:14:23   Model: clusterfinder_retrained
INFO    27/07 23:14:23   Loading model from: /home/zarul/.local/share/deepbgc/data/0.1.0/detector/clusterfinder_retrained.pkl
INFO    27/07 23:14:23   Type: DiscreteHMM
INFO    27/07 23:14:23   Version: 0.1.0
INFO    27/07 23:14:23   Timestamp: 1551449925.734045 (2019-03-01T22:18:45.734045)
INFO    27/07 23:14:23   --------------------------------------------------------------------------------
INFO    27/07 23:14:23   Model: deepbgc
INFO    27/07 23:14:23   Loading model from: /home/zarul/.local/share/deepbgc/data/0.1.0/detector/deepbgc.pkl
Using TensorFlow backend.
/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint16 = np.dtype([("quint16", np.uint16, 1)])
/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint32 = np.dtype([("qint32", np.int32, 1)])
/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:525: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  np_resource = np.dtype([("resource", np.ubyte, 1)])
/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:541: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:542: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:543: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:544: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint16 = np.dtype([("quint16", np.uint16, 1)])
/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:545: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint32 = np.dtype([("qint32", np.int32, 1)])
/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:550: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  np_resource = np.dtype([("resource", np.ubyte, 1)])
WARNING: Logging before flag parsing goes to stderr.
W0727 23:14:23.820647 140009826789184 deprecation_wrapper.py:119] From /home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:74: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.

W0727 23:14:23.829619 140009826789184 deprecation_wrapper.py:119] From /home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:517: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.

W0727 23:14:23.830600 140009826789184 deprecation_wrapper.py:119] From /home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:4138: The name tf.random_uniform is deprecated. Please use tf.random.uniform instead.

W0727 23:14:24.130435 140009826789184 deprecation_wrapper.py:119] From /home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:133: The name tf.placeholder_with_default is deprecated. Please use tf.compat.v1.placeholder_with_default instead.

W0727 23:14:24.134406 140009826789184 deprecation.py:506] From /home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:3445: calling dropout (from tensorflow.python.ops.nn_ops) with keep_prob is deprecated and will be removed in a future version.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
W0727 23:14:24.515460 140009826789184 deprecation_wrapper.py:119] From /home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:174: The name tf.get_default_session is deprecated. Please use tf.compat.v1.get_default_session instead.

I0727 23:14:24.652741 140009826789184 info.py:26] Type: KerasRNN
I0727 23:14:24.652947 140009826789184 info.py:27] Version: 0.1.0
I0727 23:14:24.653012 140009826789184 info.py:28] Timestamp: 1551305667.986168 (2019-02-28T06:14:27.986168)
I0727 23:14:24.653129 140009826789184 info.py:22] --------------------------------------------------------------------------------
I0727 23:14:24.653177 140009826789184 info.py:23] Model: clusterfinder_original
I0727 23:14:24.653223 140009826789184 wrapper.py:174] Loading model from: /home/zarul/.local/share/deepbgc/data/0.1.0/detector/clusterfinder_original.pkl
I0727 23:14:24.655262 140009826789184 info.py:26] Type: ClusterFinderHMM
I0727 23:14:24.655326 140009826789184 info.py:27] Version: 0.1.0
I0727 23:14:24.655377 140009826789184 info.py:28] Timestamp: 1551449904.5101252 (2019-03-01T22:18:24.510125)
I0727 23:14:24.655647 140009826789184 info.py:67] ================================================================================
I0727 23:14:24.655702 140009826789184 info.py:68] Available classifiers: ['product_class', 'product_activity']
I0727 23:14:24.655784 140009826789184 info.py:22] --------------------------------------------------------------------------------
I0727 23:14:24.655827 140009826789184 info.py:23] Model: product_class
I0727 23:14:24.655869 140009826789184 wrapper.py:174] Loading model from: /home/zarul/.local/share/deepbgc/data/0.1.0/classifier/product_class.pkl
/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/sklearn/base.py:306: UserWarning: Trying to unpickle estimator DecisionTreeClassifier from version 0.18.2 when using version 0.21.2. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/sklearn/base.py:306: UserWarning: Trying to unpickle estimator RandomForestClassifier from version 0.18.2 when using version 0.21.2. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
I0727 23:14:24.696886 140009826789184 info.py:26] Type: RandomForestClassifier
I0727 23:14:24.696993 140009826789184 info.py:27] Version: 0.1.0
I0727 23:14:24.697047 140009826789184 info.py:28] Timestamp: 1551781410.019103 (2019-03-05T18:23:30.019103)
I0727 23:14:24.700250 140009826789184 info.py:22] --------------------------------------------------------------------------------
I0727 23:14:24.700344 140009826789184 info.py:23] Model: product_activity
I0727 23:14:24.700432 140009826789184 wrapper.py:174] Loading model from: /home/zarul/.local/share/deepbgc/data/0.1.0/classifier/product_activity.pkl
I0727 23:14:24.710430 140009826789184 info.py:26] Type: RandomForestClassifier
I0727 23:14:24.710538 140009826789184 info.py:27] Version: 0.1.0
I0727 23:14:24.710602 140009826789184 info.py:28] Timestamp: 1551781433.886473 (2019-03-05T18:23:53.886473)
I0727 23:14:24.710908 140009826789184 info.py:78] ================================================================================
I0727 23:14:24.710979 140009826789184 info.py:80] All OK
```
I tried to run deepbgc on a file containing 25 contigs, all above 20kb but it didn't complete (I specified only deepbgc detector). There was a _tmp_ directory containing all the prodigal and pfam files. There was an empty _evaluation_ directory - not expected based on your example results notebook. Here is the last few lines of LOG.txt:

```
Detected 36 total BGCs using deepbgc model
Number of BGCs with predicted product_class: 
 no confident class: 22
 Polyketide: 8
 Saccharide: 3
 RiPP: 3
 Terpene: 1
 NRP: 1
Number of BGCs with predicted product_activity: 
 antibacterial: 26
 no confident class: 9
 cytotoxic: 1
 antifungal: 1
tuple index out of range
Traceback (most recent call last):
  File "/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/deepbgc/main.py", line 113, in main
    run(argv)
  File "/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/deepbgc/main.py", line 102, in run
    args.func.run(**args_dict)
  File "/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/deepbgc/command/pipeline.py", line 187, in run
    writer.close()
  File "/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/deepbgc/output/evaluation/pfam_score_plot.py", line 28, in close
    self.save_plot()
  File "/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/deepbgc/output/evaluation/pfam_score_plot.py", line 59, in save_plot
    axes[i].plot(x, y, lw=0.75, alpha=0.6, color=color, label=column)
  File "/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/matplotlib/axes/_axes.py", line 1666, in plot
    lines = [*self._get_lines(*args, data=data, **kwargs)]
  File "/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/matplotlib/axes/_base.py", line 225, in __call__
    yield from self._plot_args(this, kwargs)
  File "/home/zarul/Zarul/Software/anaconda3/envs/bgc/lib/python3.7/site-packages/matplotlib/axes/_base.py", line 399, in _plot_args
    ncx, ncy = x.shape[1], y.shape[1]
IndexError: tuple index out of range
================================================================================
DeepBGC failed with IndexError: tuple index out of range
================================================================================
```

Thank you.

**Comment by prihoda** [comment]
Hi Zarul, 

glad to hear about Geof's seminar and thanks for the bug report! 

Have you tried installing deepbgc using conda in the empty environment or just in your base conda environment? And did that fail? I am having trouble reproducing the error.

The TF and sklearn deprecation warnings should not impact the results, we used to specify the exact TF version in our dependencies, but then we figured users might find it limiting if they are using other tools in the same environment. Feel free to use `pip install tensorflow==1.12.0 scikit-learn==0.18.2` to avoid the deprecation warnings.

As for the error, that was probably caused by trying to plot a sequence with just a single detected pfam domain. I added new tests for that functionality and fixed the error https://github.com/Merck/deepbgc/pull/12. Please try upgrading using `pip install deepbgc==0.1.18` and running deepbgc again with the same parameters (the pfam annotation results should be reused using the tmp folder). 

Cheers,
David

**Comment by ZarulHanifah** [comment]
Hello David,

Thank you for replying. Conda kept reminding me to update from 4.7.7 to 4.7.10, so I tried to update conda but failed. So I did a new conda installation, and now deepbgc installation with conda easily works! And the plots also appear, with deepbgc 0.1.8! Everything worked simply with a new conda version (why?).

Anyways, thank you very much David for your help.

Regards,
Zarul

**Comment by prihoda** [comment]
Hi Zarul, great to hear that. I fixed the plotting bug in version `0.1.8` after you reported this issue so that's why it works now 😄About the conda issue, it's hard to see what was going on but I am glad it was resolved by updating. I am closing this issue, please let us know if you bump into anything else.
